psi4/src/psi4/libfock/snLinK.cc

```diff
@@ -0,0 +1,548 @@
+    
+    // => Part #1: Options Processing <= //
+    timer_on("snLinK: Options Processing");
+
+    // set a few Psi4-specific parameters
+    incfock_iter_ = false;
+  
+    basis_tol_ = options.get_double("SNLINK_BASIS_TOLERANCE");
+
+    pruning_scheme_ = options_.get_str("SNLINK_PRUNING_SCHEME");
+    radial_scheme_ = options_.get_str("SNLINK_RADIAL_SCHEME");
+ 
+//#if psi4_SHGSHELL_ORDERING == LIBINT_SHGSHELL_ORDERING_STANDARD
+#if psi4_SHGSHELL_ORDERING == 1 
+    is_cca_ = true;
+//#elif psi4_SHGSHELL_ORDERING == LIBINT_SHGSHELL_ORDERING_GAUSSIAN
+#elif psi4_SHGSHELL_ORDERING == 2 
+    is_cca_ = false;
+#else
+    #error "unknown value of macro psi4_SHGSHELL_ORDERING"
+#endif
+
+    format_ = Eigen::IOFormat(4, 0, ", ", "\n", "[", "]");
+    
+    // create mappings for GauXC pruning and radial scheme enums
+    auto [ pruning_scheme_map, radial_scheme_map ] = generate_enum_mappings();
+
+    // define runtime environment and execution space
+    use_gpu_ = options_.get_bool("SNLINK_USE_GPU"); 
+    auto ex = use_gpu_ ? GauXC::ExecutionSpace::Device : GauXC::ExecutionSpace::Host;  
+
+    std::unique_ptr<GauXC::RuntimeEnvironment> rt = nullptr; 
+#ifdef GAUXC_HAS_DEVICE 
+    if (use_gpu_) {
+        rt = std::make_unique<GauXC::DeviceRuntimeEnvironment>( GAUXC_MPI_CODE(MPI_COMM_WORLD,) 0.01*options_.get_int("SNLINK_GPU_MEM"));
+    } else { 
+        rt = std::make_unique<GauXC::RuntimeEnvironment>( GAUXC_MPI_CODE(MPI_COMM_WORLD) );
+    }
+#else
+        rt = std::make_unique<GauXC::RuntimeEnvironment>( GAUXC_MPI_CODE(MPI_COMM_WORLD) );
+#endif
+
+    // get maximum supported AM for this GauXC instance
+    gauxc_max_am_ = GauXC::gauxc_max_am(ex, GauXC::SupportedAlg::SNLINK);
+
+    // basis set shouldn't have higher AM shells than GauXC instance supports 
+    if (primary_->max_am() > gauxc_max_am_) {
+        std::string error_message = "Selected basis set has higher-AM shells than supported by current GauXC instance! \n";
+        error_message += "Either reduce the size of your basis set, or use a GauXC install with a higher supported AM.";
+
+        throw PSIEXCEPTION(error_message);
+    }
+
+    // cross-check supported L2 and GauXC AMs
+    // yoinking L2 AM from psi4/driver/p4util/util.py's libint2_configuration() function 
+    auto p4util = py::module::import("psi4").attr("p4util"); // from psi4 import p4util
+    auto libint2_configuration = p4util.attr("libint2_configuration")(); // p4util.libint2_configuration
+    int l2_max_am = libint2_configuration["eri"][py::int_(0)].cast<int>();
+
+    if (l2_max_am != gauxc_max_am_) {
+        std::string error_message = "Libint2 and GauXC do not have equivalent maximum AM support!";
+
+        throw PSIEXCEPTION(error_message);
+    }
+
+    // set whether we force use of cartesian coordinates or not
+    // this is required for GPU execution when using spherical harmonic basis sets
+    auto force_cartesian = options_.get_bool("SNLINK_FORCE_CARTESIAN"); 
```

Comment:
>Ok, I see now those are info, not error. So I guess it's only CPU and sph basis sets >that both values of force_cartesian=T/F have an effect?

Correct. `SNLINK_FORCE_CARTESIAN` obviously can't have an effect on Cartesian basis sets; and `SNLINK_FORCE_CARTESIAN` is forcibly set to `T` for GPU calculations.

`SNLINK_FORCE_CARTESIAN` has an effect on CPU+Sph basis sets specifically to allow debugging of the Spherical-to-Cartesian transform process, where the end result should be the same regardless of `SNLINK_FORCE_CARTESIAN`'s value. `test_compositejk.py` currently utilizes this debugging functionality.

>https://github.com/psi4/psi4/blob/master/psi4/src/export_mints.cc#L97-L106 >shelltype has the result of the negotiations for sph/cart btwn the basis set itself and >what the user wants it computed as. Could the result of that be used instead of ?>requiring a separate snlink_force_cartesian be set by user?

Perhaps. Technically, though, the user never _has_ to set `SNLINK_FORCE_CARTESIAN` to get a reasonable value, since it is forcibly set in all cases except CPU+Sph basis sets (where it defaults to `false` anyway). `SNLINK_FORCE_CARTESIAN` also allows for debugging/testing the Sph-to-Cart transformation handling in the interface code.